### ISSUE TITLE
Implement shadow culling based on line of sight

### ISSUE THREAD

**Issue by julealgon** [diasurgical/devilutionX#780]
One feature I always wished was present in D1 is line-of-sight based visibility of the map. One of the few things that still are immersion breaking to me is the fact that the player can see the entire map even if the character is not able to actually see behind walls.

I believe this has been implemented to some extent in Diablo 2 but I feel like it would be really cool to see it in Diablo 1 as well. It would accentuate it's terror atmosphere even more and making for some interesting surprises when playing.

For folks who are potentially not aware of what I'm describing here, this is a nice interactive example:
https://www.redblobgames.com/articles/visibility/

**Comment by qndel** [comment]
![image](https://user-images.githubusercontent.com/14297035/91673572-03eed800-eb35-11ea-839a-a67a5dae0dcf.png)

**Comment by julealgon** [comment]
Hey @qndel I was actually going to mention you here because I was looking at the per-pixel lighting you did. Both things at the same time would be amazing.

Do you know if the entire code responsible for lighting and the game grid are all fully translated to C/C++ already? I noticed there was a huge amount of ASM code in the repo a few months ago. Not sure what's the status now.

I could attempt something if the code is at a high enough level of abstraction right now. I'm quite proficient at C# and know a bit of C++, but if I had to work on raw assembly then there is no hope 😆

**Comment by qndel** [comment]
yes, all the asm got rewritten to C/C++, there's no asm in devilutionx

**Comment by strich** [comment]
@julealgon don't suppose there is still interest from you in doing this at this stage? It's a great idea!

**Comment by julealgon** [comment]
> [@julealgon](https://github.com/julealgon) don't suppose there is still interest from you in doing this at this stage? It's a great idea!

@strich Interest? Absolutely... this is one of the main things I want to eventually do (at least in my mod). The problem is motivation and real life stuff... right now, it's just not realistic for me to say I'll do it as I'd be lying (to you and to myself).

However, if I end up doing it, it will likely not be for DevilutionX anymore but for my own version (which I haven't really started yet, shamefully):
- https://github.com/julealgon/DiabloCore/issues/26

If someone wants to implement this for DevilutionX I'd say go for it.

**Comment by strich** [comment]
I'm definitely tempted to tackle more modern colored pixel lights. I saw you were the one who gave that a go a few years ago - Did you stall out and get stuck?

**Comment by StephenCWills** [comment]
@kphoenix137 did this in his mod, Lord of Terror. Might be good to talk to him about a possible DevilutionX implementation. Though, given that he hasn't already submitted a PR for DevilutionX, he probably doesn't consider this to be a "clean" implementation, and likely considers it to be challenging to make it so.

Anyway, here's how his looks.

https://github.com/user-attachments/assets/50c03865-acad-46c1-8ae7-9d07be2b46da

**Comment by strich** [comment]
Holy shit! That's awesome. What is that dude doing not bringing that back into a PR. :(

**Comment by julealgon** [comment]
> I'm definitely tempted to tackle more modern colored pixel lights. I saw you were the one who gave that a go a few years ago - Did you stall out and get stuck?

@strich I think you may be referring to the "pixel light" mod/idea? That was @qndel 's. I think the main issue there was it required hardware lighting so wouldn't be fully compatible with DevX's current restrictions but that might be out of date.
- https://github.com/diasurgical/DevilutionX/pull/559

If you were referring to something else and I'm just confusing it, let me know.

> @kphoenix137 did this in his mod, Lord of Terror.

Had no idea kphoenix worked on that! Looks very cool, could be a good basis for a more polished implementation.

**Comment by romanstingler** [comment]
I am working on an implementation for this project (not checked LoT). 

https://github.com/user-attachments/assets/480095c3-2895-4b48-a148-94107a0d4daf
https://github.com/user-attachments/assets/9cdd8b2d-09ff-4c6e-a1ff-616ea908ac55

I have to check a few more things and as soon as I have a version where I think it is good for testing I will let you know.
PS: it is a debug build so it is slow(low-fps)
